### PR TITLE
[Connector API] Support cleaning up sync jobs when deleting a connector

### PR DESCRIPTION
**Pull request by jedrazb**
### Changes
- Support cleaning up associated sync jobs in sync jobs index when deleting a connector, this is controlled by `delete_sync_jobs` URL param for the delete request
- This flag is set to `true` by default, example of how to opt out: `DELETE _connector/{id}?delete_sync_jobs=false`
- Added unit tests and yaml tests
- Since this uses `delete_by_query` under the hood I needed to link `Reindex` plugin for unit tests in `build.gradle`
- Added small workaround to handle template mappings in unit tests, this allows to defined "simplified" mappings to test functions that rely on particular field type 
  - We have several tests that `AwaitsFix` for the issue linked to missing template mappings - I will follow up in next PR to check if all of the tests are fixed now

### CODE REVIEW

File: x-pack/plugin/ent-search/build.gradle [hunk 1]
```diff
@@ -30,6 +30,7 @@ dependencies {
   testImplementation(testArtifact(project(xpackModule('core'))))
   testImplementation project(":test:framework")
   testImplementation(project(':modules:lang-mustache'))
+  testImplementation(project(':modules:reindex'))
```
Comment (jedrazb) [inline x-pack/plugin/ent-search/build.gradle:33]: It seems to be needed for `ESSingleNodeTestCase`  to register `Reindex` plugin that exposes `delete_by_query` action

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/320_connector_delete.yml [hunk 2]
```diff
@@ -26,6 +26,49 @@ setup:
...
+
+  - match: { acknowledged: true }
+
+
+  - do:
+      connector_sync_job.list:
+        connector_id: test-connector-to-delete
+
+  - match: { count: 0 }
+
```
Comment (timgrein) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/320_connector_delete.yml:71]: Maybe an explicit integration test testing `delete_sync_jobs=false` would also make sense

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java [hunk 3]
```diff
@@ -253,12 +254,13 @@ public void getConnector(String connectorId, ActionListener<ConnectorSearchResul
     }
 
     /**
-     * Deletes the {@link Connector} in the underlying index.
+     * Deletes the {@link Connector} and the related sync jobs in the underlying index.
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:257]:
```suggestion
     * Deletes the {@link Connector} and the related instances of {@link ConnectorSyncJob} in the underlying index.
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java [hunk 4]
```diff
@@ -253,12 +254,13 @@ public void getConnector(String connectorId, ActionListener<ConnectorSearchResul
...
      *
-     * @param connectorId The id of the connector object.
-     * @param listener    The action listener to invoke on response/failure.
+     * @param connectorId          The id of the connector object.
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:259]:
```suggestion
     * @param connectorId          The id of the {@link Connector}.
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java [hunk 5]
```diff
@@ -253,12 +254,13 @@ public void getConnector(String connectorId, ActionListener<ConnectorSearchResul
...
+     * @param shouldDeleteSyncJobs The flag indicating if connector sync jobs should also be deleted.
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorIndexService.java:260]:
```suggestion
     * @param shouldDeleteSyncJobs The flag indicating if {@link ConnectorSyncJob} should also be deleted.
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java [hunk 6]
```diff
@@ -35,7 +35,11 @@ public List<Route> routes() {
 
     @Override
     protected RestChannelConsumer prepareRequest(RestRequest restRequest, NodeClient client) throws IOException {
-        DeleteConnectorAction.Request request = new DeleteConnectorAction.Request(restRequest.param("connector_id"));
+
+        String connectorId = restRequest.param("connector_id");
+        boolean shouldDeleteSyncJobs = restRequest.paramAsBoolean("delete_sync_jobs", true);
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java:40]: Maybe double-confirm the default value with product keeping for example auditing purposes in mind, where you never want to loose sync jobs

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java [hunk 7]
```diff
@@ -586,6 +593,37 @@ public void updateConnectorSyncJobError(String connectorSyncJobId, String error,
         }
     }
 
+    /**
+     * Deletes all {@link ConnectorSyncJob} documents that match a specific connector id in the underlying index.
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:597]:
```suggestion
     * Deletes all {@link ConnectorSyncJob} documents that match a specific {@link Connector} id in the underlying index.
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java [hunk 8]
```diff
@@ -586,6 +593,37 @@ public void updateConnectorSyncJobError(String connectorSyncJobId, String error,
...
+     * Gracefully handles non-existent sync job index.
+     *
+     * @param connectorId The id of the connector to match in the sync job documents.
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/syncjob/ConnectorSyncJobIndexService.java:600]:
```suggestion
     * @param connectorId The id of the {@link Connector} to match in the sync job documents.
```

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTestUtils.java [hunk 9]
```diff
@@ -46,11 +47,53 @@
...
+import static org.elasticsearch.xpack.application.connector.ConnectorTemplateRegistry.CONNECTOR_SYNC_JOBS_INDEX_NAME_PATTERN;
+import static org.elasticsearch.xpack.application.connector.ConnectorTemplateRegistry.CONNECTOR_SYNC_JOBS_TEMPLATE_NAME;
+import static org.elasticsearch.xpack.application.connector.ConnectorTemplateRegistry.CONNECTOR_TEMPLATE_NAME;
 
 public final class ConnectorTestUtils {
 
     public static final String NULL_STRING = null;
 
+    /**
+     * Registers index templates for connectors and sync jobs with essential field mappings. This method
```
Comment (timgrein) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorTestUtils.java:60]:
```suggestion
     * Registers index templates for instances of {@link Connector} and {@link ConnectorSyncJob} with essential field mappings. This method
```

**Review by timgrein — APPROVED** [review]
Just some minor comments, LGTM

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java (hunk 6 above)
Comment (wangch079) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java:40]: And also for backward-compatibility

Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/action/RestDeleteConnectorAction.java:40]: @danajuratoni  suggested to follow default Kibana behaviour. 
Defaulting this to false is the default in Kibana, also it's backward compatible, so I will follow that approach.

**Pull request merged**